### PR TITLE
Migration To CircleCI 2.0

### PR DESCRIPTION
**Pull request by marcomorain**
### Description of the Change

Migrate `atom/atom` from CircleCI 1.0 to CircleCI 2.0 for macOS builds.

### Alternate Designs

No alternate designs were considered.

### Why Should This Be In Core?

This is an update to CI config.

### Benefits

At CircleCI we have announced that we are going to [EOL CircleCI 1.0 in August](https://circleci.com/blog/sunsetting-1-0/). In advance of this, I've been taking a look at some of the bigger open-source projects using CircleCI 1.0 for macOS, and helping them to migrate to 2.0 in plenty of time. In turn, this will help us to better manage capacity of our build fleet.

The build should be faster on 2.0. Our macOS runs on the same physical hardware for 1.0 and 2.0, but the our network has been much improved for our 2.0 fleet. 2.0 also provides much more fine-grained control over caching, which often speeds up builds because we can skip uploading caches that have not changed.

### Possible Drawbacks

The atom core team might not be familiar with the CircleCI 2.0 config syntax.

### Verification Process

I have run some builds on a fork - https://circleci.com/gh/marcomorain/atom-1 - the build is not yet green, which I assume is because I do not have the code-signing certificates and other settings which are stored in private env-vars in https://circleci.com/gh/atom/atom

### Applicable Issues

The build is not yet green.

### CODE REVIEW

**Comment by marcomorain** [comment]
One question - is it required to pin the node version?

The old build did this:

```
dependencies:
   pre:
     - curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.31.3/install.sh | bash
     - nvm install 6.9.4
     - nvm use 6.9.4
     - npm install -g npm@5.3.0
```

On CircleCI, `xcode: 9.2.0` is pinned to Node 8.9.1 (yarn 1.3.2). This won't change unless you change xcode version, to a new image.

**Comment by marcomorain** [comment]
Does anyone have time to review this please?

**Comment by Arcanemagus** [comment]
> One question - is it required to pin the node version?

The Node.js version was likely pinned for compatibility (testing builds work on the minimum version) and build reliability (new version of Node.js not breaking the build).

Atom will occasionally lag behind the released version of Node.js in what is supported for builds. _Currently_ at least the latest versions work fine for bootstrapping the build, but this hasn't always been the case. The same applies to the npm version.

I also notice that you have removed the workarounds that prevent the screensaver from running. Is it disabled already on the newer images?

----

167 test failures looks like a _major_ problem, virtually all of the ones I see are timeouts... could this be due to the screensaver issue above?

**Comment by maxbrunsfeld** [comment]
> One question - is it required to pin the node version?

I think node 8.9.1 should be fine. @marcomorain Is the version of npm guaranteed to be stable as well? If not, we'll probably want to continue at least installing npm explicitly.

**Comment by maxbrunsfeld** [comment]
> the build is not yet green

> could this be due to the screensaver issue above?

I added back the screensaver workarounds from our Circle 1.0 config, and there are still hundreds of timeouts. @marcomorain Do you have any insight into the differences from 1.0 to 2.0 that might be causing these problems?

**Comment by daviwil** [comment]
Hey @marcomorain, thanks a lot for getting this started!  Is there anything we can help with to get the tests passing again?

**Comment by daviwil** [comment]
I'm going to see if I can get this PR unblocked so that we can get moved over to Circle 2.0 by the cutoff date.

**Comment by marcomorain** [comment]
Sorry I meant to get back to this but I never did.

File: package.json [hunk 1]
```diff
@@ -39,7 +39,7 @@
     "fs-plus": "^3.0.1",
     "fstream": "0.1.24",
     "fuzzaldrin": "^2.1",
-    "git-utils": "5.4.0",
+    "git-utils": "5.2.1",
```
Comment (maxbrunsfeld) [inline package.json]: See https://github.com/libgit2/libgit2/issues/4740

**Comment by maxbrunsfeld** [comment]
@marcomorain Thanks for following up on this. I think we identified the problem.

In Atom's test suite, we're hitting a [`libgit2` bug](https://github.com/libgit2/libgit2/issues/4740) where certain computations are accidentally scanning the entire history of the git repository. It looks like this wasn't a problem on Circle 1.0 because Circle 1.0 performed some kind of a shallow clone of the repository, whereas Circle 2.0 [clones the entire history](https://discuss.circleci.com/t/checkout-depth/18699) by default?

After downgrading our libgit2 wrapper, our build is passing with the exception of two known test failures. 💚

**Comment by marcomorain** [comment]
Oh that’s good news. Any idea why the build on 2.0 exposed the issue, but not 1.0?

Maybe the dependencies were cached already on 1.0, from before the libgit regression?

**Comment by marcomorain** [comment]
I’ll clean up the new config file tomorrow. 

Since we support workflows in CircleCI 2.0, we can split chunks of the build out to different containers. Are there any parts of the build and test scripts that can be run independently to parallelise things?

**Comment by jasonrudolph** [comment]
> After downgrading our libgit2 wrapper, our build is passing with the exception of two known test failures, which will pass if this PR is merged. 💚

Just a quick update: Those two test failures are resolved in https://github.com/atom/atom/pull/17752. Merging `master` into this branch should get those tests passing on this branch.

**Comment by daviwil** [comment]
Hey @marcomorain, I can't seem to restart the CircleCI build on my own, but the failures there are known test flakes on macOS so I think they'll go away with another run.

**Comment by jasonrudolph** [comment]
@marcomorain Thanks for your work on this. :bow:

Given that we're getting close to the August 31 EOL date for CircleCI 1.0, we're in a bit of a crunch to make sure we continue having a working CI solution for Atom's macOS builds. Since we recently implemented cross-platform builds (macOS, Linux, and Windows) on VSTS for Atom Nightly Releases (#17445, #17538), it feels like the pragmatic choice is to use the already-working VSTS builds as our definitive macOS CI solution going forward. With that in mind, I'm going to close this pull request, but please know that we appreciate your work on this migration, and we're grateful for all the many benefits that CircleCI has brought to Atom over the years. 💟

**Comment by julesvicious** [comment]
So I should be Izze ?

Julesvicious 

On Aug 6, 2018, at 11:19 AM, Jason Rudolph <notifications@github.com> wrote:

@marcomorain Thanks for your work on this. 🙇

Given that we're getting close to the August 31 EOL date for CircleCI 1.0, we're in a bit of a crunch to make sure we continue having a working CI solution for Atom's macOS builds. Since we recently implemented cross-platform builds (macOS, Linux, and Windows) on VSTS for Atom Nightly Releases (#17445, #17538), it feels like the pragmatic choice is to use the already-working VSTS builds as our definitive macOS CI solution going forward. With that in mind, I'm going to close this pull request, but please know that we appreciate your work on this migration, and we're grateful for all the many benefits that CircleCI has brought to Atom over the years. 💟

**Comment by marcomorain** [comment]
Thanks for the kind note @jasonrudolph 👍

**Pull request closed without merging**